rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+---
+feature: demote_darwin_support_tier
+start-date: 2021-10-27
+author: piegames
+co-authors: many
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Move x86_64-darwin from a Tier 2 platform to Tier 3, as described in [RFC 0046](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md).
+
+# Motivation
+[motivation]: #motivation
+
+Darwin is constantly adding additional maintainer burden. Especially, it does not live up to the requirements:
+
+- "A lot of packages built by Hydra, full ofBorg support."
+    - There is only one ofBorg builder. It can take several hours or days for an ofBorg build to finish for Darwin, if it finishes at all.
+    - Hydra builders a greatly understaffed.
+    - If Apple follows through on [their announcement to discontinue](https://www.businessinsider.com/apple-macbook-pro-discontinued) x86-based devices, the situation won't improve due to a physical lack of hardware.
+- "Most packages work, credible ambition to reach Tier 1 coverage at some point."
+    - Tier 1 support is far out of reach.
+    - We lack expertise in the greater community to support Darwin.
+- "Some ordinary packages are channel blockers on Hydra."
+    - Due to Hydra being understaffed, this is a problem (see above).
+
+We are not only lacking CI machines, but also maintainers that can dedicate their time to work on fixing broken Darwin packages.
+
+## Exemplary collection of issues with Darwin
+
+The issues that are constantly encountered in relation to the Darwin platform have a genuine negative impact on the `nixpkgs` development experience. People are forced to deal with these problems even though they have no hardware to test on, lack the proper expertise, and are reluctant when it comes to investing additional time and energy to patch for the failing platform. Demoting Darwin to Tier 3 will allow more developers to refocus their attention on better supported platforms, care less about Darwin issues, and push that burden onto the Darwin maintainers.
```

Comment:
I think this is the major issue with darwin support in nixpkgs, and it's not captured in how the current support tiers in rfc 46 are formulated.

The support tiers are currently summarised as
1. **If** the platform achieves the required tooling and package coverage
2. **then** it can have the impact on nixpkgs

but in the case of darwin the impact on nixpkgs as a whole is so big, it sounds like we don't want to give it tier 2 impact, regardless of whether it actually achieves the current tier 2 tooling and package coverage.

I think I agree with that, but if that's the case, let's update the tiers so that it is clear what the requirements for each tier are. For example if we think not being able to spin up a VM for the platform on all major OSes is a problem, let's list that out as a requirement.

This might have been brought up in rfc 46 too because I see:
> It is recommended to provide a derivation to test the software on this platform
> (e.g. a Qemu-based derivation with all the necessary scripts).
> As it is impossible to provide a legal testing setup for a Tier-2 platform
> (macOS), this requirement is not strictly mandatory for Tier-3 tooling.

Please correct me if I'm wrong, I was not involved in rfc 46 so I might be missing some context there and misinterpreted it.

Reply:
RFC 46 got deeply rewritten in the process, so I did try to limit its scope, as well as the _change_ it brought to Nixpkgs (and it says explicitly that movement between Tiers 1-2-3 is always a policy discussion not a purely technical one, that's also for a reason).

I support a public resolution of «if update breaks Darwin, follow #88 with the expectation that the fix will be prepared or not by people personally  using Nix on macOS, but the timer starts from the mention of the relevant Darwin-specific team», modulo a low-effort way to get the «measure the impact» data for Darwin with zero knowledge or care about Darwin (well, if Darwin capacity at OfBorg becomes truly plentiful — beyond simply reasonably sufficient for current use — maybe poking it with a subset of rebuild report would become an OK option).